### PR TITLE
Respect WP_REDIS_FLUSH_TIMEOUT

### LINKED ISSUE #508: Yoast timeouts
There have been many reports about Yoast using group flushing. Maybe those `EVAL` calls need their own timeout.

```
[01-Mar-2024 06:15:30 UTC] RedisException: socket error on read socket in /home/www/cc.cz/subdomains/www/wp-content/object-cache.php:1826
Stack trace:
#0 /home/www/cc.cz/subdomains/www/wp-content/object-cache.php(1826): Redis->eval()
#1 /home/www/cc.cz/subdomains/www/wp-content/object-cache.php(1726): WP_Object_Cache->{closure}()
#2 /home/www/cc.cz/subdomains/www/wp-content/object-cache.php(162): WP_Object_Cache->flush_group()
#3 /home/www/cc.cz/subdomains/www/wp-content/plugins/wordpress-seo/src/builders/indexable-link-builder.php(710): wp_cache_flush_group()
#4 /home/www/cc.cz/subdomains/www/wp-content/plugins/wordpress-seo/src/builders/indexable-link-builder.php(623): Yoast\WP\SEO\Builders\Indexable_Link_Builder->update_incoming_links_for_related_indexables()
#5 /home/www/cc.cz/subdomains/www/wp-content/plugins/wordpress-seo/src/builders/indexable-link-builder.php(136): Yoast\WP\SEO\Builders\Indexable_Link_Builder->update_related_indexables()
#6 /home/www/cc.cz/subdomains/www/wp-content/plugins/wordpress-seo/src/integrations/watchers/indexable-post-watcher.php(209): Yoast\WP\SEO\Builders\Indexable_Link_Builder->build()
#7 /home/www/cc.cz/subdomains/www/wp-includes/class-wp-hook.php(326): Yoast\WP\SEO\Integrations\Watchers\Indexable_Post_Watcher->build_indexable()
#8 /home/www/cc.cz/subdomains/www/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters()
#9 /home/www/cc.cz/subdomains/www/wp-includes/plugin.php(517): WP_Hook->do_action()
#10 /home/www/cc.cz/subdomains/www/wp-includes/post.php(4771): do_action()
#11 /home/www/cc.cz/subdomains/www/wp-includes/post.php(4862): wp_insert_post()
#12 /home/www/cc.cz/subdomains/www/wp-admin/includes/post.php(453): wp_update_post()
#13 /home/www/cc.cz/subdomains/www/wp-admin/post.php(227): edit_post()
#14 {main}
```

### LINKED ISSUE #510: RedisException: socket error on read socket when Publish a post (Wordpress SEO enabled)
Whenever I want to publish a post, the error `socket error on read socket` present itself. Screenshot below is the error (with Query Monitor enabled).

I wonder if this is Yoast SEO bug, or something else.

## Description

I want to publish a post. When I clicked the "Publish" button, the page produce an error `socket error on read socket`. The error log is as below:

```
2024/03/07 15:34:55 [error] 5942#5942: *967964 FastCGI sent in stderr: "PHP message: RedisException: socket error on read socket in /home/website/web/app/object-cache.php:1830
Stack trace:
#0 /home/website/web/app/object-cache.php(1830): Redis->eval()
#1 /home/website/web/app/object-cache.php(1730): WP_Object_Cache->{closure}()
#2 /home/website/web/app/object-cache.php(162): WP_Object_Cache->flush_group()
#3 /home/website/web/app/plugins/wordpress-seo/src/builders/indexable-link-builder.php(710): wp_cache_flush_group()
#4 /home/website/web/app/plugins/wordpress-seo/src/builders/indexable-link-builder.php(623): Yoast\WP\SEO\Builders\Indexable_Link_Builder->update_incoming_links_for_related_indexables()
#5 /home/website/web/app/plugins/wordpress-seo/src/builders/indexable-link-builder.php(136): Yoast\WP\SEO\Builders\Indexable_Link_Builder->update_related_indexables()
#6 /home/website/web/app/plugins/wordpress-seo/src/integrations/watchers/indexable-post-watcher.php(20" while reading response header from upstream, client: 118.101.126.13, server: thevocket.com, request: "POST /wp/wp-admin/post.php HTTP/2.0", upstream: "fastcgi://unix:/run/php/php8.2-fpm.sock:", host: "thevocket.com", referrer: "https://thevocket.com/wp/wp-admin/post.php?post=485476&action=edit"
```
thevocket com_wp_wp-admin_post php

## Expected Behavior

I expect the post will save, publish and back to the editing page.

## Actual Behavior

Halted by error `socket error on read socket` and the post did not redirect back to the editing page.

## Possible Fix

Not sure. I have tried to implement all suggestion installation configuration, including the redis.conf `lazyfree-lazy-user-flush` as `yes` as suggest here https://github.com/rhubarbgroup/redis-cache/pull/490#issuecomment-1795948671

## Steps to Reproduce

1. Edit or Create new Posts
2. Save as Draft
3. Press "Publish"

## Additional context

I was trying to publish an article.

## Environment

* WordPress Setup: roots/bedrock
* Plugin version: 2.5.1
* PHP version: 8.2.15
* WordPress version: 6.4.3
* Yoast SEO plugin: 22.1
* Redis Server: 7.2.4
* PhpRedis: 6.0.2
* Predis: 2.1.2
* Nginx: 1.25.4

## Constants
* Config::define('WP_REDIS_HOST', '127.0.0.1');
* Config::define('WP_REDIS_PORT', 6379);
* Config::define('WP_REDIS_PREFIX', 'my-site');
* Config::define('WP_REDIS_DATABASE', 0);

### LINKED ISSUE #514: Command `_MASTERS` is not a registered Redis command when enable in cli
## Description
Hi, your worst nightmare here... :upside_down_face:
According to https://github.com/rhubarbgroup/redis-cache/issues/512, I´m trying to enable Redis from command cli, and I get the following error:
```bash
[apache@myserver /opt/apache/htdocs/wordpress-site]$ php wp-cli.phar redis enable
Predis\ClientException: Command `_MASTERS` is not a registered Redis command. in /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/dependencies/predis/predis/src/Command/Factory.php:67
Stack trace:
#0 /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/dependencies/predis/predis/src/Client.php(313): Predis\Command\Factory->create()
#1 /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/dependencies/predis/predis/src/Client.php(304): Predis\Client->createCommand()
#2 /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/includes/class-predis.php(164): Predis\Client->__call()
#3 /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/includes/class-predis.php(198): Rhubarb\RedisCache\Predis->flush()
#4 /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/includes/cli/class-commands.php(176): Rhubarb\RedisCache\Predis->flushOrFail()
#5 /opt/apache/htdocs/wordpress-site/wp-content/plugins/redis-cache/includes/cli/class-commands.php(63): Rhubarb\RedisCache\CLI\Commands->flush_redis()
#6 [internal function]: Rhubarb\RedisCache\CLI\Commands->enable()
#7 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/WP_CLI/Dispatcher/CommandFactory.php(100): call_user_func()
#8 [internal function]: WP_CLI\Dispatcher\CommandFactory::WP_CLI\Dispatcher\{closure}()
#9 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/WP_CLI/Dispatcher/Subcommand.php(491): call_user_func()
#10 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(431): WP_CLI\Dispatcher\Subcommand->invoke()
#11 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(454): WP_CLI\Runner->run_command()
#12 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1269): WP_CLI\Runner->run_command_and_exit()
#13 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(28): WP_CLI\Runner->start()
#14 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/bootstrap.php(83): WP_CLI\Bootstrap\LaunchRunner->process()
#15 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/vendor/wp-cli/wp-cli/php/wp-cli.php(32): WP_CLI\bootstrap()
#16 phar:///opt/apache/htdocs/wordpress-site/wp-cli.phar/php/boot-phar.php(20): include('phar:///opt/htt...')
#17 /opt/apache/htdocs/wordpress-site/wp-cli.phar(4): include('phar:///opt/htt...')
#18 {main}
Error: Object cache could not be enabled. Redis server is unreachable: Command `_MASTERS` is not a registered Redis command.
```
If I enable this from Wordpress admin area, is enabled correctly.

## Expected Behavior
Enable Redis from command cli

## Actual Behavior
Exception

## Steps to Reproduce

1. Install new wordpress site and add the plugin
2. Activate the plugin
3. Enable fthe plugin from command cli

## Environment

* Plugin version: **2.5.1**
* PHP version: **7.4.30**
* WordPress version: **6.4.3**
* Configuration in `wp-config.php`
```php
//REDIS CACHE
define('WP_CACHE', true);
define('WP_CACHE_KEY_SALT', 'wordpress_site_2024');
define('WP_REDIS_SCHEME', 'tcp');
define('WP_REDIS_HOST', 'redisint3.dev');
define('WP_REDIS_PORT', '6379');

define('WP_REDIS_CLUSTER', [
        'tcp://redisint1.dev:6379?alias=redisint1',
        'tcp://redisint2.dev:6379?alias=redisint2',
        'tcp://redisint3.dev:6379?alias=redisint3'
]);
```

### PR DESCRIPTION
**Pull request by yatsukhnenko**
If set use value of `WP_REDIS_FLUSH_TIMEOUT` when flush cache
TODO: get/set timeout helpers should be added to Predis

Closes #510
Closes #508
Closes #514

### CODE REVIEW

**Review by tillkruss — CHANGES_REQUESTED** [review]
I think this can work, but I think we need to refactor this a little bit. We can leave `lua_flush_closure()` and all the other lua script functions as they are.

Can we add a new function, something like execute_lua_script() that will:

1. take the lua closure
2. Handle clusters and instance connections
3. Handle settings timeouts for Predis / PhpRedis

**Comment by yatsukhnenko** [comment]
@tillkruss I tried to handle case when Redis cluster is used with predis and it looks like there is a lack of functionality in predis
```php
Fatal error: Uncaught Predis\NotSupportedException: Cannot use 'EVAL' with redis-cluster. in /var/www/html/wp-content/plugins/redis-cache/dependencies/predis/predis/src/Connection/Cluster/RedisCluster.php:353
```

**Comment by yatsukhnenko** [comment]
Nevermind, there is no error when call `eval` directly on node

File: includes/object-cache.php [hunk 1]
```diff
@@ -1590,6 +1590,92 @@ public function flush_runtime() {
...
+     * @return array|false  Returns array on success, false on failure
+     */
+    protected function execute_lua_script( $script ) {
+        $results = [];
+
+        if ( defined( 'WP_REDIS_CLUSTER' ) ) {
+            return $this->execute_lua_script_on_cluster( $script ) 
+        }
+
+        if ( defined( 'WP_REDIS_FLUSH_TIMEOUT' ) ) {
```
Comment (tillkruss) [inline includes/object-cache.php:1605]: Let's default to `5s` if `WP_REDIS_FLUSH_TIMEOUT` is not set.

**Review by tillkruss — CHANGES_REQUESTED** [review]

**Pull request merged**